Use getTimeRotatingLogger for WMCore REST

d-ylee opened the pull request:
Fixes/Is a part of #11922 

#### Status
In development | not-tested

#### Description
Should set up a Python TimedRotatingFileHandler to replace the usage of `rotatelogs`

#### Is it backward compatible (if not, which system it affects?)
MAYBE
Should only be used when `|rotatelogs ...` is not used during deployment

#### Related PRs
https://github.com/dmwm/CMSKubernetes/pull/1452
https://github.com/dmwm/CMSKubernetes/pull/1604

#### External dependencies / deployment changes
Potentially remove the use of `rotatelogs`.

d-ylee committed:
Removed rotatelogs setup in RESTDaemon

Use getTimeRotatingLogger for WMCore REST

Added date to filename for MyTimedRotatingFileHandler if it doesn't have it

Added usage for wmc-httpd LOG-FILE arg

Renamed MyTimedRotatingFileHandler to WMTimedRotatingFileHandler

Simplified WMTimedRotatingFileHandler to use namer function

Remove unused code and docstrings

Fixed service log rotation

Add QueueHandler for cherrypy, remove redirecting logging

QueueHandler for cherrypy logs

Added a queue listener thread that aggregates all logs that are handled
with the WMTimedRotatingLogger

Removed redirecting logging to a subprocess or a file using file
descriptors

vkuznet reviewed (approved):
Dennis, changes are fine but I don't know how `getTimeRotatingLogger` works and will rely on you that you tested it and confirmed that it works as expected.

d-ylee commented:
@amaltaro I was testing this and it was rotating the logs, but not retaining them or adding the date. I found out that the date was originally provided in `rotatelogs` when calling `wmc-httpd`:

With `rotatelogs`:
https://github.com/dmwm/CMSKubernetes/blob/326de33e3769e2f66e088c91093d3179a8d1cc12/docker/pypi/reqmgr2ms-unmerged/run.sh#L145

Without `rotatelogs`:
https://github.com/dmwm/CMSKubernetes/blob/a6dfdee30f4a2cabc308663f2f3a9b215e08e991/docker/pypi/reqmgr2ms-unmerged/run.sh#L145

Now, I changed `MyTimedRotatingFileHandler` to add the YearMonthDate string when initializing the logger. I'm not sure if we want to inject the date 7via Python, or if we want to 100% replicate how we use `rotatelogs` by passing what we had originally in `rotatelogs` and replace `%Y%m%d` string in the `MyTimedRotatingFileHandler.__init__` function with a `todayStr`.

Example
In `run.sh`:
```
wmc-httpd -r -d $STATEDIR -l "$LOGDIR/$srv-%Y%m%d-`hostname -s`.log 86400" $CFGFILE
```

In `WMLogging.MyTimedRotatingFileHandler.__init__`
```
 def __init__(self, logName, interval, backupCount):
        # Add todayStr to log name, like with rotatelogs
        todayStr = date.today().strftime("%Y%m%d")
        filename = logName.replace('%Y%m%d', todayStr)
        super(MyTimedRotatingFileHandler, self).__init__(filename, when=interval,
                                                         backupCount=backupCount)
```

d-ylee commented:
To add a note:
This function doesn't do what the docstring says:
https://github.com/dmwm/WMCore/blob/dd7835203b951857de3a544c0a5dc8ed50d121c7/src/python/WMCore/WMLogging.py#L68-L85

The new date string is replaced, not added. So if the string doesn't contain yesterdayStr, then the string isn't added and the baseName remains the same.

vkuznet reviewed (commented):
Dennis, the changes are fine but I would suggest to improve description and add to it exact syntax how to use this when we invoke python program., in other words, so far we use ``wmc-httpd -r -d $STATEDIR -l "|rotatelogs $LOGDIR/$srv-%Y%m%d-`hostname -s`.log 86400" $CFGFILE`` and it would be beneficial to have an example how we will call similar daemon (or python command) with new logger.

d-ylee committed:
Added test for WMCore.WMLogging.MyTimedRotatingFileHandler

Updated WMLogging test to test a quick log rotation

Fix some pylint errors in WMLogging_t.py

Replace existing log rotation test

amaltaro commented on src/python/WMCore/REST/Main.py:
```diff
@@ -619,5 +625,8 @@ def main():
             else:
```
Perhaps the option above in the `if` statement should be removed and we define this option to be in the format of `-l my_logfile.log` only. In other words, `|`, `rotatelogs` and `86400` (integer with time in seconds) are no longer an acceptable option. What do you think?

amaltaro reviewed (changes requested):
@d-ylee I left a few comments/questions along the code.
I also wanted to point out to this line: https://github.com/dmwm/WMCore/blob/6228aa0268f7cab8e36cb9d0698c1d1ddb4de4e3/src/python/WMCore/REST/Main.py#L319

which IMO should be removed with this PR. In other words, I would say the log file name option has to be as simple as `-l filename.log` or `-l filename_date.log` and that is it. No need to accept pipe, rotatelogs and time in secs. Any thoughts?

d-ylee commented:
@amaltaro If we remove the lines in RESTDaemon as you suggested, should we also remove this:
https://github.com/dmwm/WMCore/blob/6228aa0268f7cab8e36cb9d0698c1d1ddb4de4e3/src/python/WMCore/REST/Main.py#L410-L413

I think this was used to setup streaming to `rotatelogs`.

The other thing is that the line you referenced:
https://github.com/dmwm/WMCore/blob/6228aa0268f7cab8e36cb9d0698c1d1ddb4de4e3/src/python/WMCore/REST/Main.py#L319

...gets overwritten if we provide `-l`:
https://github.com/dmwm/WMCore/blob/6228aa0268f7cab8e36cb9d0698c1d1ddb4de4e3/src/python/WMCore/REST/Main.py#L622-L629

So the change will have the RESTDaemon default to logging to a file.

amaltaro commented:
@d-ylee yes, I think you are correct about the `Popen` and I am inclined to say that we should completely remove that, as we no longer expect (accept?) a list type parameter for the logfile.

Given that this line:
```
self.logfile = ["rotatelogs", "%s/%s-%%Y%%m%%d.log" % (self.statedir, self.appname), "86400"] 
```
was being used as default and get overwritten by the `-l` option. How about we keep this default value, but with the following difference:
```
self.logfile = "%s/%s-%%Y%%m%%d.log".format(self.statedir, self.appname, todayStr) 
```
or perhaps simply:
```
self.logfile = "%s/%s.log".format(self.statedir, self.appname) 
```
?

For now, I'd suggest to keep it in a different commit, just in case we regret this change :)

d-ylee commented:
test this please

amaltaro commented:
@d-ylee if we merge these changes in, doesn't it mean we also have to update how we execute the services to stop using `rotatelogs`?

If so, then I think we need to update the `dmwm-base` image as well, see:
```
docker/pypi/dmwm-base/run.sh:wmc-httpd -r -d $STATEDIR -l "|rotatelogs $LOGDIR/$srv-%Y%m%d-`hostname -s`.log 86400" $CFGFILE
docker/pypi/dmwm-base/manage:    wmc-httpd -r -d $STATEDIR -l "|rotatelogs $LOGDIR/$srv-%Y%m%d-`hostname -s`.log 86400" $CFGFILE
docker/pypi/reqmgr2ms-unmerged/run.sh:wmc-httpd -r -d $STATEDIR -l "|rotatelogs $LOGDIR/$srv-%Y%m%d-`hostname -s`.log 86400" $CFGFILE
```

To summarize the changes that I see required for this:
1) finish up with this WMCore pull request
2) make a new CMSKubernetes PR to update `dmwm-base` image
3) create a new `dmwm-base` tag and adopt it in `reqmgr2ms-unmerged`
4) update `reqmgr2ms-unmerged` docker image to Alma9 plus Python 3.12
5) build a "private" MSUnmerged docker image
6) test the new image. If we are happy with the behavior, we then need to update all of the other WMCore Dockerfiles to adopt the new `dmwm-base` tag

and at this point, we have fully deprecated Apache `rotatelogs`. What do you think? Did I miss anything?

d-ylee commented:
@amaltaro I think this would be the case. We would probably need to move `dmwm-base` away from Debian.

An alternative could be to stay in Debian, but I think our goal is to move to Alma9. The initial prototype you made here: https://github.com/dmwm/CMSKubernetes/pull/1452 could be a good starting point, but we also should be careful not to break any existing image building.

amaltaro commented:
@d-ylee I thought we could move away from `rotatelogs` package while keeping the same current base image (python 3.8 debian). I am overlooking something?

d-ylee commented:
@amaltaro Yes I think we can do this. I think I was probably talking about the bigger picture, but I think we can follow the steps mentioned here: https://github.com/dmwm/WMCore/pull/11955#issuecomment-2770558301

d-ylee commented:
@amaltaro Is there anything we should add to this pull request?

amaltaro commented:
Once we have this change in, we can likely remove the dependency on `apache2-utils` from the dmwm-base image:
https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/dmwm-base/Dockerfile#L6C82-L6C95

I don't know if anything is used from this package, but rotatelogs is:
```
_reqmgr2@reqmgr2-f44b4f784-4d9j5:/data$ dpkg -S rotatelogs
apache2-utils: /usr/share/man/man8/rotatelogs.8.gz
apache2-utils: /usr/bin/rotatelogs
```

d-ylee commented:
@amaltaro I created a PR here for removing `apache2-utils`

https://github.com/dmwm/CMSKubernetes/pull/1610

d-ylee commented:
After doing the testing for [CMSKubernetes#1604](https://github.com/dmwm/CMSKubernetes/pull/1604), there are some issues during the rollover period of the log file, with the file not actually being renamed. I am doing more further investigation.

d-ylee commented:
@amaltaro While doing some investigation today, I noticed that MSUnmerged is creating 3 log files:
1. `heartbeatMonitor`: Created from the CherryPyPeriodicTask heartbeat monitor task; defined in config-unmerged.py (affected by this PR)
2. `reqmgr2ms-unmerged-<timestamp>.log`: Created from the main daemon (affected by this PR)
3. `reqmgr2ms-unmerged.log`: Created from CherryPy REST API Logger; contents are the result of logging via `cherrpy.log()`, as seen here: https://github.com/dmwm/WMCore/blob/e9e236bb4c63813e3cdfa2d5dd3092c3c1d4cec8/src/python/WMCore/REST/Main.py#L265

[1] and [2] both utilize the WMTimedRotatingFileHandler that is being worked on in this PR.

[3] on the other hand seems to be something separate. Should the CherryPy RESTApi logs be in the same file as [2]?
